oscore-cf/src/main/java/org/eclipse/californium/oscore/OSCoreCtx.java

```diff
@@ -0,0 +1,673 @@
+
+		try {
+			this.recipient_key = deriveKey(this.common_master_secret, this.common_master_salt, this.key_length, digest,
+					info.EncodeToBytes());
+		} catch (CoseException e) {
+			LOGGER.severe(e.getMessage());
+			throw new OSException (e.getMessage());
+		}
+
+		// Derive common_iv
+		info = CBORObject.NewArray();
+		info.Add(new byte[0]);
+		info.Add(this.context_id);
+		info.Add(this.common_alg.AsCBOR());
+		info.Add(CBORObject.FromObject("IV"));
+		info.Add(this.iv_length);
+
+		LOGGER.finest("Info common IV: " + bytesToHex(info.EncodeToBytes()));
+
+		try {
+			this.common_iv = deriveKey(this.common_master_secret, this.common_master_salt, this.iv_length, digest,
+					info.EncodeToBytes());
+		} catch (CoseException e) {
+			LOGGER.severe(e.getMessage());
+			throw new OSException (e.getMessage());
+		}
+		
+		System.out.println("Common_IV");
+		for(byte b: this.common_iv){
+			System.out.print(Integer.toHexString(b & 0xff) + " ");
+		}
+		System.out.println();
+		
+	}
+
+	/**
+	 * Overrides hasCode to provide a functional implementation for this class.
+	 */
+	@Override
+	public int hashCode() {
+		byte[] c = new byte[sender_id.length + recipient_id.length];
+		System.arraycopy(sender_id, 0, c, 0, sender_id.length);
+		System.arraycopy(recipient_id, 0, c, sender_id.length, recipient_id.length);
+		return ByteBuffer.wrap(c).hashCode();
+	}
+
+	/**
+	 * Overrides equals to provide a functional implementation for this class.
+	 */
+	@Override
+	public boolean equals(Object o) {
+		if (!(o instanceof OSCoreCtx)) {
+			return false;
+		}
+		OSCoreCtx other = (OSCoreCtx) o;
+		return other.sender_id.equals(sender_id) && other.recipient_id.equals(recipient_id);
+	}
+
+	/**
+	 * @return get the sender key
+	 */
+	public byte[] getSenderKey() {
+		return sender_key;
+	}
+
+	/**
+	 * @return get the recipient key
+	 */
+	public byte[] getRecipientKey() {
+		return recipient_key;
+	}
+
+	/**
+	 * @return get the encryption algorithm
+	 */
+	public AlgorithmID getAlg() {
+		return this.common_alg;
+	}
+
+	/**
+	 * @return get the sender sequence number
+	 */
+	public synchronized int getSenderSeq() {
+		return sender_seq;
+	}
+
+	/**
+	 * @return get the receiver sequence number
+	 */
+	public synchronized int getReceiverSeq() {
+		return recipient_seq;
+	}
+
+	/**
+	 * @return the tag of the last block processed with this context
+	 */
+	public byte[] getLastBlockTag() {
+		return last_block_tag;
+	}
+
+	/**
+	 * @return the sender's identifier
+	 */
+	public byte[] getSenderId() {
+		return sender_id;
+	}
+
+	/**
+	 * @return the repipient's identifier
+	 */
+	public byte[] getRecipientId() {
+		return recipient_id;
+	}
+
+	/**
+	 * Compares value to the current notification number
+	 * 
+	 * @param value
+	 * @return true if value is largare than notification_number, otherwise
+	 *         false
+	 */
+	public boolean isNewNotificationNumbervalid(int value) {
+		return value > notification_number;
```

Comment:
Related to RFC7641? Should it not comply with https://tools.ietf.org/html/rfc7641#section-4.4 ?

Reply:
We will see, how it is used, when it get's used.
I'm not sure, if it's easier to remove it until it gets really used,

Reply:
I feel like the PR will be leaner without them and removing them will have zero impact on this PR and the next one.